Derived units

ericman314 opened the pull request:
Here's a first attempt at derived units. (See issue #13.) I've also added new unit tests. Check out the updated examples. I think temperatures might not work altogether correctly yet, so will need to add some unit tests to that effect.

Also, any recommendations on an editor for .md files? I'm assuming you'd like some help eventually updating those.

FSMaxB commented:
I can recommend mdcharm for markdown editing.

josdejong commented:
Thanks Eric, this looks really good already!! Thanks for making serious work of the unit tests and examples directly, this is really important. In the end we indeed need to update the docs too. I use WebStorm which is great but not for markdown. For editing markdown I sometimes use this online editor: http://jbt.github.io/markdown-editor/.

A few ideas/remarks:
- There is quite some logic (with regexp's) for parsing a derived unit. This would be needed when `Unit.js` was a standalone library, but in this case of math.js we have the expression parser already. This means that for typical usage, you would do `math.eval('2 km/h')`, where the derived unit will always be created via operations multiply, divide, power. How do you think about keeping the constructor `Unit.js` simple in this regard and just rely on the expression parser for constructing a derived unit?
-  Maybe operations such as `multiply` and `divide` can become simpler and faster when the array Unit.units is always sorted?

josdejong commented on examples/basic_usage.js:
```diff
@@ -34,6 +34,7 @@ console.log();
 console.log('mixed use of data types');
 print(math.add(4, [5, 6]));                   // number + Array, [9, 10]
 print(math.multiply(math.unit('5 mm'), 3));   // Unit * number,  15 mm
+print(math.multiply(math.unit('8 ft'), math.unit('20 lbf')).to('ft lbf'));   // Unit * number,  15 mm
```
It's indeed nice to showcase as much as possible in this very small example, but I would prefer to keep this example as it is, short and simple, and go nuts in the units.js example.

josdejong commented:
This is going to be great, I'm really looking forward to this :)

ericman314 commented:
Jos, great suggestions. I do like the idea of using the parser for constructing a unit. One of the advantages of doing the parsing in the constructor was that if you did <code>var a = new Unit("2 N m")</code>, the new unit would contain the original N and m units so that the unit is formatted as <code>2 N m</code>. But when writing <code>var a = math.eval("2 N m")</code>, the <code>multiply</code> function is called by the parser, which obliterates the N and formats the new unit as <code>2 kg m^2 / s^2</code>. But I agree that we should leverage the existing parser.

Rewriting <code>multiply</code> and <code>divide</code> is definitely a good idea, and maybe we could do it so it "remembered" as many of the existing units as possible, or perhaps substituted a better unit (like <code>J</code> in place of <code>N m</code>). 

The opposite approach is to use something like a number array that contains the powers for each dimension (length, time, mass, etc.) so that it is very fast to multiply and divide, but then the unit would not remember any of its own history.

There are two directions we could take:
<ol>
<li>Keep the units very flexible (as in a list of arbitrary units with powers) and let <code>multiply</code>, <code>divide</code>, and <code>pow</code> intelligently combine units and decide on the best form.</li>
<li>Make the internal structure of the units very simple, such as only storing SI units or having a fixed-size array, and let the formatter decide on the best units to display. In this case, we would need some kind of history to remember what units this unit was created from.
</ol>

Another symptom of the present dilemma is that doing <code>math.eval('2 m/s to km/h')</code> fails with <code>Error: Cannot convert to a unit with a value</code>, since <code>km/h</code> is evaluated by the parser as <code>0.27777777777778 m / s</code>. Interestingly, <code>math.eval('2 m/s to "km/h"')</code> (with <code>km/h</code> in quotes) <i>does</i> work (returns <code>7.2 km / h</code>), by virtue of the parsing in the Unit constructor.

Any thoughts? I can probably work on it this week.

ericman314 commented:
I'm having trouble calling the parser from the Unit constructor. I've tried a bunch of variations of the following:

```javascript
lib/type/unit/Unit.js:

var math = require('../../../index');

function factory (type, config, load, typed) {

  function Unit(value, name) {
    var u = math.eval(name);
  }

}
```
Fails with:
```
[TypeError: Object #<Object> has no method 'eval']
```
Am I missing something obvious?

josdejong commented:
You can do something like this:

```js
function factory (type, config, load, typed) {
  var eval = load(require('../../expression/function/eval'));

  // or, a bit closer to the metal, use `parse(expr).eval()`
  var parse = load(require('../../expression/parse'));

  function Unit(value, name) {
    var u = eval(name);
  }
}
```

You can't directly access the `math` namespace containing all these functions. In v2 we changed to a modular structure where you explicitely have to load/require the functions that you need. This allows to load/bundle only the functions of math.js that you actually need ([see docs for more info](http://mathjs.org/docs/custom_loading_and_bundling.html)). Please let me know if there are other things unclear.

ericman314 commented:
Thanks. I tried both of those but it fails with:
```
RangeError: Maximum call stack size exceeded
```
Seems like a circular dependency. I think I can cheat by doing ```exports.math = true``` (BigNumber, for example), but is there a better way to do it?

josdejong commented:
argh, I should have thought of that. There is indeed a circular reference: parse.js depends on SymbolNode.js, which depends on Unit.js, so when referencing parse.js form Unit.js we have a problem.

I don't think it's safe to use the math namespace here, it's easily possible that someone bundles the functions and data types and excludes the expression parser. I think there are two options:

- Your initial solution: write a complete parser for derived units in Unit.js itself. Requires extra code but it may be acceptable.

- Don't include a parser for derived units in Unit.js itself (used as `math.type.Unit.parse`), but only add this to the constructor function `./lib/type/unit/function/unit.js`. That way we can safely use parse.js without introducing a circular reference.

- The only part of `Unit.js` needed by Symbol.js is the function `Unit.isValuelessUnit`, which in turn loops over the array with `Unit.UNITS`. We could split this function and the objects with units (UNITS, BASE_UNITS, PREFIXES, etc) from the Unit.js function into a separate helper file `definitions.js`, and reference this helper file from both Units.js and parse.js. Something like: 

  ```js
  // lib/type/unit/definitions.js
  function factory (type, config, load, typed) {
      return {
        UNITS: {...},
          PREFIXES: {....},
      // ...
      findUnit: function (str) {...},   // the current _findUnit
      isValuelessUnit: function (str) {...}
    }
  });

  // no exports.name needed, we don't want to expose this in the math namespace
  exports.factory = factory;
  ```

  Both `Unit.js` and `parse.js` can now use `var definitions = load(require('./definitions'))` and use `definitions.UNITS`, `definitions.isValuelessUnit`, etc.

I don't like the trickiness of this circular reference. From that point of view we should not use `parse.js` from `Unit.js`. `Unit.js` is lower in the hierarchy than `parse.js` (data types are lowest, and functions operate on data types), so it doesn't feel right to have a dependency to `parse.js` from `Unit.js`. Making Unit.js standalone, having it's own mini-parser is "safest", but it costs more code. That may be acceptable, it also depends on how much code we are talking about.

What do you think?

ericman314 commented:
I think we should write our own parser in ```Units.js```. The format for units is very simple. I can do it character-by-character without regex's if you prefer. It should be faster than using ```parse.js```.

josdejong commented:
I agree. Ok then let's go for that.

ericman314 commented:
I think it's reached a semi-decent state.

- `Unit` has its own simple parser now without regex's.
- Added a dimensions array (one element each for `LENGTH`, `MASS`, etc.). This makes multiplication/etc. very fast. Unit lists are concatenated in the result so that the unit "remembers" what units it was created from.
- The `to` operator works now.
- `Multiply`, `divide`, and `pow` do not simplify units immediately; instead, units are simplified before they are formatted for output. This lazy evaluation hopefully will speed things up.
- Valueless units are never simplified (this enables things like `math.eval('25 m/s to km/h')` to work.)
- Added a few "unit systems" like `si`, `cgs`, and `us` that the user can choose from that can help the unit simplification function to choose the right units. Haven't documented this function in the docs yet.
- There is also an `auto` unit system that remembers the most recently used units, and uses those in the unit simplification function when possible.
- Added some energy, pressure, and power units.
- Updated unit tests, examples, and docs.

pearswj commented:
@ericman314, awesome work. Threw my own simple test* at it and the results are as expected :smile:

```
# Simply Supported Beam

This calculates the maximum moment for a simply supported beam.

Two inputs are required:
* The **load per metre** along the beam:
$w = 5 kN m^-1$
* The **length** of the beam:
$l = 2 m$

The **maximum moment** can now be calculated.

$M_max = (w*l^2)/8$
```

<img width="912" alt="screen shot 2015-08-04 at 10 14 34" src="https://cloud.githubusercontent.com/assets/121068/9057457/003d5eba-3a92-11e5-82c8-fcf04d5984e5.png">

*a _very early_ prototype for an open tool for evaluating and typesetting engineering calculations

josdejong commented:
wow! This is awesome!! From what I've tested it works like a charm. Really cool :)

One set of units that we still miss is electricity, we could add that too. From one of my [previous applications](http://www.speqmath.com/usersguide/?page=units_overview.html):

```
A (Ampere)	
C (Coulomb)	s*A
W (Watt)	J/s
V (Volt)	W/A
Ohm (Resistance)	V/A
F (Farad)	C/V
Wb (Weber)	J/A
T (Tesla)	Wb/m2
H (Henry)	Wb/A
S (Siemens)	1/Ohm
eV (ElectronVolt)	1.6021765314e-19 J
```
